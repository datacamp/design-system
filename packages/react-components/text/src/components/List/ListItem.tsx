import { css } from '@emotion/core';
import React, { ReactElement, ReactNode } from 'react';

const listItemStyle = css({
  lineHeight: '1.5em',
  marginTop: 8,
});

export interface ListItemProps {
  children: ReactNode;
  className?: string;
}

const ListItem = ({ children, className }: ListItemProps): ReactElement => {
  return (
    <li className={className} css={listItemStyle}>
      {children}
    </li>
  );
};

ListItem.displayName = 'List.Item';

export default ListItem;
