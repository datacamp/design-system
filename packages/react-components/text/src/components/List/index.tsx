import tokens from '@datacamp/waffles-tokens';
import {
  computeDataAttributes,
  ssrSafeNotFirstChildSelector,
} from '@datacamp/waffles-utils';
import { css } from '@emotion/react';
import React from 'react';

import baseStyle from '../../baseStyle';

import ListItem from './ListItem';

interface ListProps {
  /**
   * The content to display. Should only be List.Item components
   */
  children: React.ReactNode;
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
  /**
   * Blah Blah Blah docs
   */
  ordered?: boolean;
}

const listStyle = css(baseStyle, {
  // add padding when nested
  '& &': {
    paddingLeft: tokens.spacing.medium,
  },
  listStylePosition: 'inside',
  margin: 0,
  padding: 0,
  [ssrSafeNotFirstChildSelector]: {
    marginTop: tokens.spacing.xsmall,
  },
});

const List = (props: ListProps): JSX.Element => {
  const { children, className, dataAttributes, ordered = false } = props;

  const parsedDataAttributes = computeDataAttributes(dataAttributes);
  const Element = ordered ? 'ol' : 'ul';

  return (
    <Element className={className} css={listStyle} {...parsedDataAttributes}>
      {children}
    </Element>
  );
};

List.Item = ListItem;

export default List;
