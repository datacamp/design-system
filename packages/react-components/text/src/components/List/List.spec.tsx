import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import React from 'react';

import List from '.';

describe('List', () => {
  it('renders an unordered list', async () => {
    const { container } = await axeRender(
      <List className="test-class">
        <List.Item>top level item 1</List.Item>
        <List.Item>top level item 2</List.Item>
        <List.Item>
          nested items
          <List>
            <List.Item>nested item 1</List.Item>
            <List.Item>nested item 2</List.Item>
          </List>
        </List.Item>
      </List>
    );
    expect(container.firstChild).toHaveClass('test-class');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an ordered list', async () => {
    const { container } = await axeRender(
      <List className="test-class">
        <List.Item>top level item 1</List.Item>
        <List.Item>top level item 2</List.Item>
        <List.Item>
          nested items
          <List>
            <List.Item>nested item 1</List.Item>
            <List.Item>nested item 2</List.Item>
          </List>
        </List.Item>
      </List>
    );
    expect(container.firstChild).toHaveClass('test-class');
    expect(container.firstChild).toMatchSnapshot();
  });
});
