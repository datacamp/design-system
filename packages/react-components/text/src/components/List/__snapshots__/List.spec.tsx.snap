// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders an ordered list 1`] = `
.emotion-0 {
  color: rgb(5, 25, 45);
  font-family: Studio-Feixen-Sans,Arial;
  font-style: normal;
  list-style-position: inside;
  margin: 0;
  padding: 0;
}

.emotion-0 .emotion-0 {
  padding-left: 16px;
}

*:not(style)~.emotion-0 {
  margin-top: 16px;
}

.emotion-1 {
  line-height: 1.5em;
  margin-top: 8px;
}

<ul
  class="test-class emotion-0"
>
  <li
    class="emotion-1"
  >
    top level item 1
  </li>
  <li
    class="emotion-1"
  >
    top level item 2
  </li>
  <li
    class="emotion-1"
  >
    nested items
    <ul
      class="emotion-0"
    >
      <li
        class="emotion-1"
      >
        nested item 1
      </li>
      <li
        class="emotion-1"
      >
        nested item 2
      </li>
    </ul>
  </li>
</ul>
`;

exports[`List renders an unordered list 1`] = `
.emotion-0 {
  color: rgb(5, 25, 45);
  font-family: Studio-Feixen-Sans,Arial;
  font-style: normal;
  list-style-position: inside;
  margin: 0;
  padding: 0;
}

.emotion-0 .emotion-0 {
  padding-left: 16px;
}

*:not(style)~.emotion-0 {
  margin-top: 16px;
}

.emotion-1 {
  line-height: 1.5em;
  margin-top: 8px;
}

<ul
  class="test-class emotion-0"
>
  <li
    class="emotion-1"
  >
    top level item 1
  </li>
  <li
    class="emotion-1"
  >
    top level item 2
  </li>
  <li
    class="emotion-1"
  >
    nested items
    <ul
      class="emotion-0"
    >
      <li
        class="emotion-1"
      >
        nested item 1
      </li>
      <li
        class="emotion-1"
      >
        nested item 2
      </li>
    </ul>
  </li>
</ul>
`;
